### PR TITLE
add support for progressive + pal60 video mode forcing (for Launch Disc)

### PR DESCRIPTION
**Pull request by pyorot**
as discussed in #376; this is my sketch for the solution:

1. determine render mode from system settings (`CONF_GetProgressiveScan`, `VIDEO_HaveComponentCable` and `CONF_GetEuRGB60`).
2. set render mode register within the `VIDEO_Configure` wrapper function, determined from render mode.
3. ~~remove gamecube-disc-launch `SYS_SetVideoMode` call cos undocumented and not used in non-gc-disc paths (note: i'm unsure if this function supports the value 5 for PAL60 and don't want it to undo the register setting in step 2).~~
4. ~~add video mode setting to title-launch path (i didn't spot earlier it wasn't being called because the called function has "Title" in its name so i assumed it would be lol)~~

edit: 3 needs to stay and 4 doesn't fix issue #376... in fact, it rebreaks it after having been fixed by a different change (#379) lol. 1 and 2 seem like they would make disc loading more robust and correct (e.g. starting pal discs in progressive) tho so i've left them up here

this still doesn't support stuff like mpal and so isn't totally libogc-compliant but that's why we have the libogc dev on-hand to sort it out fully :3

not tested, ~~not compiled, not checked by a language server even (i wasn't expecting to be doing any c++ :p)~~

### CODE REVIEW

**Comment by pyorot** [comment]
i may be interested in compiling this and running it on my boot2-hacked wii at some point, so any tips for setting up libogc and everything would be appreciated

**Comment by DacoTaco** [comment]
did you try this change? did this fix your issue?

sorry i didn't get back to you sooner, my life is a bit busy atm

**Comment by pyorot** [comment]
no worries. i haven't tested or compiled the code yet, and may give it a shot sometime soon, but wanted to open it up for comments in the meantime. i'd have to get the dev setup working and that kind of thing

i think my issue is caused more by the old code being functionally handicapped rather than wrong. it always sets interlaced for cross-region so even routes that do force video, like wii disc loads, should theoretically do that – e.g. ntsc-u disc in a pal wii will always start in 480i for example. as a european, i only have my backups in wad form to test unfort :)

**Comment by DacoTaco** [comment]
i would suggest to first test this change then, to compile you need devkitPRO's devkitPPC and libogc and that should be it (its all under then `wii-dev` package collection).
i can already see a few issues that i remember were critical, but i wont be doing any code review until its confirmed working for your case tbh

**Comment by pyorot** [comment]
> i would suggest to first test this change then, to compile you need devkitPRO's devkitPPC and libogc and that should be it (its all under then `wii-dev` package collection). i can already see a few issues that i remember were critical, but i wont be doing any code review until its confirmed working for your case tbh

fair, it was at least very easy to set up the toolchain so thanks for enabling that

i am testing rn, current patch doesn’t work and i’ve traced the bug to the title region detection logic. SetVideoModeForTitle isn’t being called at all. there’s a very similar bug in usbloader gx it seems, where “detect disc” starts 480i but “force ntsc” (NOT “force ntsc 480p”) starts 480p despite the code setting 480i lol. to be continued…

**Comment by pyorot** [comment]
i got stuck tryna figure out the mystery in usbloader gx that made Disc Default boot 480i yet both Force 480i and Force 480p boot 480p (lol), so i decided to isolate out the patches that fixed the blocking issue of broken region detection logic, and accidentally fixed the whole thing :p. new patchset up at #379

File: src/priiloader/source/titles.cpp [hunk 1]
```diff
@@ -404,15 +404,24 @@ s8 SetVideoModeForTitle(u32 lowerTitleId)
...
+				rmodeNew = &TVEurgb60Hz480Prog;
+				gprintf("PAL60 480p");
+			} else if (confPAL60) {		// 480i60
+				rmodeNew = &TVEurgb60Hz480IntDf
+				gprintf("PAL60 480i");
+			} else {					// 576i50 (incompatible with S-Video cables!)
+				rmodeNew = &TVPal528IntDf;
+				gprintf("PAL50 576i");
+			}
 			videoMode = SYS_VIDEO_PAL;
```
[the remark this answers is not in the document]
Comment (pyorot) [reply src/priiloader/source/titles.cpp:415]: oh nice one, ty. what does `SYS_Set_Video` do? i'll replace my patch removing it with some comments

**Comment by DacoTaco** [comment]
is this still applicable? this will need some rebasing though :)

**Comment by pyorot** [comment]
> is this still applicable? this will need some rebasing though :)

the first commit by itself is just tidiness/correctness i guess, always setting that video mode address to the pending viTVMode Format (0–5) rather than just 0 or 1. the second i feel like should fix cross-region discs always booting interlaced but i've not verified that actually happens. i do know someone i could ask to check it tho

how does it look to you just statically?

**Comment by DacoTaco** [comment]
> > is this still applicable? this will need some rebasing though :)
> 
> the first commit by itself is just tidiness/correctness i guess, always setting that video mode address to the pending viTVMode Format (0–5) rather than just 0 or 1. the second i feel like should fix cross-region discs always booting interlaced but i've not verified that actually happens. i do know someone i could ask to check it tho
> 
> how does it look to you just statically?

much better now with the rebase, the PR now contains a single change which is a lot easier to process. it looks ok, but does it actually work? are you able to test it for channels and discs?

**Comment by pyorot** [comment]
> much better now with the rebase, the PR now contains a single change which is a lot easier to process.

yeah it’s a funny thing with hacking, you have no idea what will work so you pile up 6 patches in one draft pr then turns out u only need some of them. i at least try to do one commit per change so one can click thru the commits. then i cherry-pick them out at the end

> it looks ok, but does it actually work? are you able to test it for channels and discs?

there’ve been developments with wii swiss booter last few days so i haven’t yet reached out to a guy i know maintaining guides for these things. i don’t have any cross-region discs but i was thinking of sending him the latest priiloader dev build to try booting ntsc-j gamecube SMS on a pal wii, which should answer the question of if this pr is needed

with channels, all that code is not even run cos setting the video mode seems to force interlaced for reasons i can’t fathom. so ig it’s just about if it helps with discs

btw, is there a way to test priiloader w/o installing it? i’ve been enjoying doing “make run” with wiiload lately heh

**Comment by DacoTaco** [comment]
> > much better now with the rebase, the PR now contains a single change which is a lot easier to process.
> 
> yeah it’s a funny thing with hacking, you have no idea what will work so you pile up 6 patches in one draft pr then turns out u only need some of them. i at least try to do one commit per change so one can click thru the commits. then i cherry-pick them out at the end
> 
> > it looks ok, but does it actually work? are you able to test it for channels and discs?
> 
> there’ve been developments with wii swiss booter last few days so i haven’t yet reached out to a guy i know maintaining guides for these things. i don’t have any cross-region discs but i was thinking of sending him the latest priiloader dev build to try booting ntsc-j gamecube SMS on a pal wii, which should answer the question of if this pr is needed
> 
> with channels, all that code is not even run cos setting the video mode (whether with this patch or not) seems to force interlaced for reasons i can’t fathom. so ig it’s just about if it helps with discs
> 
> btw, is there a way to test priiloader w/o installing it? i’ve been enjoying doing “make run” with wiiload lately heh

i would send it for testing ye. im looking to either fixup & merge this PR or close it because it doesnt bring anything to the table. 

as for testing, sure. make run should run priiloader just fine. keep in mind that if you have autoboot settings it will run those as well so make sure to disable those when testing!
when run from HBC some access rights errors might pop up (like title names not being read), but overal it should run fine

**Comment by pyorot** [comment]
nice one, i messaged him to drop by if he gets a sec to test so we’ll see :)

**Comment by pyorot** [comment]
> keep in mind that if you have autoboot settings it will run those as well so make sure to disable those when testing!

BRO i have asked so many stupid questions this week i’m sorry looool it was 100% this

File: src/priiloader/source/titles.cpp [hunk 2]
```diff
@@ -414,17 +422,24 @@ s8 SetVideoModeForTitle(u32 lowerTitleId)
...
+			if (confProg) {          // 480p60
+				rmodeNew = &TVNtsc480Prog;
+				gprintf("NTSC 480p");
+			} else {                 // 480i60
+				rmodeNew = &TVNtsc480IntDf;
+				gprintf("NTSC 480i");
+			}
 			videoMode = SYS_VIDEO_NTSC;
-			vidmode = &TVNtsc480IntDf;
+			gprintf("NTSC");
```
Comment (DacoTaco) [inline src/priiloader/source/titles.cpp:433]: this gprintf can be removed, as the if/else will gprintf already :)

**Review by DacoTaco — CHANGES_REQUESTED** [review]
minor change before merge :)

**Comment by DacoTaco** [comment]
can you also rebase your branch to release, as i want to release this under 10.1

**Comment by DacoTaco** [comment]
to be merged in main -> next update is 0.11 because previous patch related to this was merged in main

**Pull request merged**